src/components/workflow/workflow.service.ts

```diff
@@ -0,0 +1,871 @@
+  }
+
+  // creates a relationship from one state to another
+  // later we can create an abstracted function that creates and attaches a state to another state
+  // addPossibleStateToState
+  async addPossibleState(session: ISession, input: PossibleState): Promise<void>{
+    try{
+      const result = await this.db
+        .query()
+        .match([
+          [
+            ...matchSession(session),
+            relation('in', '', 'member'),
+            node('sg', 'SecurityGroup')
+          ],
+          [
+            node('fromState', 'State', {
+              id: input.fromStateId
+            })
+          ],
+          [
+            node('toState', 'State', {
+              id: input.toStateId
+            })
+          ]
+        ])
+        .merge([
+          node('fromState'),
+          relation('out', '', 'nextPossibleState', {
+            active: true
+          }),
+          node('toState')
+        ])
+        .first();
+      
+      if ( !result ) {
+        throw new NotFoundException('could not make correct query result');
+      }
+
+    } catch (e) {
+      this.logger.warning('failed to add possible state to state', {
+        exception: e
+      });
+      throw e;
+    }
+  }
+
+  // removePossibleStateFromState
+  async removePossibleState(session: ISession, input: PossibleState): Promise<void>{
+    try{
+      await this.db
+        .query()
+        .match([
+          [
+            ...matchSession(session),
+            relation('in', '', 'member'),
+            node('sg', 'SecurityGroup'),
+            relation('in', '', 'securityGroup', {
+              active: true
+            }),
+            node('fromState', 'State', {
+              id: input.fromStateId
+            }),
+            relation('out', 'rel', 'nextPossibleState', {
+              active: true
+            }),
+            node('toState', 'State', {
+              id: input.toStateId
+            })
+          ],
+        ])
+        .detachDelete('rel')
+        .run();
+
+    } catch (e) {
+      this.logger.warning('failed to remove possible state from state', {
+        exception: e
+      });
+      throw e;
+    }
+  }
+
+  // there will be more than one required field relationship between a state node and a base node.
+  // this is so each required field can be queried without inspecting the property name in app code.
+  // addRequiredFieldToState
+  async addField(session: ISession, input: RequiredField): Promise<void>{
+    try{
+      const field = await this.db
+        .query()
+        .match([
+          [
+            node('token', 'Token', {
+              active: true,
+              value: session.token
+            }),
+            relation('in', '', 'token', {
+              active: true
+            }),
+            node('user'),
+            relation('in', '', 'admin', {
+              active: true
+            }),
+            node('baseNode'),
+            relation('out', '', `${input.propertyName}`),
```

Comment:
the relationship type should be `requiredProperty` and the `value` property should be the `${input.propertyName}`

Reply:
relation('out', '', `requiredProperty`, {value: ${input.propertyName}),

or something like that.